### PR TITLE
Inline Trips & Metrics Event Processing

### PR DESCRIPTION
**Pull request by rymarczy**
Our Trips and Metrics processing was added after our revised RT events processing pipeline was implemented in our development environment. For that reason, the original Trip and Metrics processing implementation worked by processing batches of records so it could catch up to processing current RT records. 

Now that events process and Trips/Metrics processing is in sync, it is possible to move the processing of Trips & Metrics processing in-line with RT events. 

In this updated flow, new Trips records will be loaded based on the RT events produced in `l0_gtfs_rt_events.py`. This paves the way for removing all Trips related information from the `VehicleEvents` schema and reducing our RDS storage size. 

Metrics processing is also updated to only process the `start_date` values existing in the RT events dataframe. The records still must be processed in a batch format, since current metrics rely on comparing values to RT events from outside of the current dataframe window. 

Asana Task: https://app.asana.com/0/1204389777951216/1204119636952108

### CODE REVIEW

File: .github/workflows/continuous_integration.yaml [hunk 1]
```diff
@@ -18,7 +18,7 @@ defaults:
 jobs:
   setup:
     name: Setup
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
```
Comment (rymarczy) [inline .github/workflows/continuous_integration.yaml:21]: bump CI workflow Runners to fix errors

File: python_src/src/lamp_py/migrations/versions/performance_manager/2e9d81949da8_trips_and_metrics_processing_updates.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        sa.Column("fk_static_timestamp", sa.Integer(), nullable=False),
+    )
+    op.create_foreign_key(
+        "vehicle_trips_fk_static_timestamp_fkey",
+        "vehicle_trips",
+        "static_feed_info",
+        ["fk_static_timestamp"],
+        ["timestamp"],
+    )
+    # ### end Alembic commands ###
```
Comment (rymarczy) [inline python_src/src/lamp_py/migrations/versions/performance_manager/2e9d81949da8_trips_and_metrics_processing_updates.py:44]: RDS schema migrations:
- make `trunk_route_id` in `vehcile_trips` nullable for new trips loading process
- make `stop_count` in `vehcile_trips` nullable for new trips loading process
- add `fk_static_timestamp` to `vehicle_trips` as foreign key from `timestamp` in `static_feed_info`

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py [hunk 3]
```diff
@@ -123,18 +125,9 @@ def combine_events(
...
-    """
-    # TODO(zap / ryan) - figure out how 2 compute these. the update
-    # logic will also need to be updated.
-    return events
-
-
 def upload_to_database(
     events: pandas.DataFrame, db_manager: DatabaseManager
-) -> None:
+) -> pandas.DataFrame:
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:130]: Return the events from `upload_to_database` for use in the trips / metrics loading pipeline.

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py [hunk 4]
```diff
@@ -255,13 +244,16 @@ def upload_to_database(
     events["vp_move_timestamp"] = events["vp_move_timestamp"].astype("Int64")
     events["vp_stop_timestamp"] = events["vp_stop_timestamp"].astype("Int64")
     events["tu_stop_timestamp"] = events["tu_stop_timestamp"].astype("Int64")
+    events["vp_stop_db"] = events["vp_stop_db"].astype("Int64")
+    events["vp_move_db"] = events["vp_move_db"].astype("Int64")
     events["pk_id"] = events["pk_id"].astype("Int64")
+    events = events.fillna(numpy.nan).replace([numpy.nan], [None])
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:250]: Dataframe value cleanup to ensure `do_update` and `do_insert` masks behave properly

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py [hunk 5]
```diff
@@ -255,13 +244,16 @@ def upload_to_database(
...
 
     process_logger = ProcessLogger("gtfs_rt.update_events")
     process_logger.log_start()
 
     # update all of the events that have pk_ids and new timestamps.
-    update_mask = (events["pk_id"].notna()) & (
+    events["do_update"] = (events["pk_id"].notna()) & (
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:256]: save `do_update` mask for later use in trips processing

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py [hunk 6]
```diff
@@ -294,10 +285,10 @@ def upload_to_database(
     process_logger.log_start()
 
     # any event that doesn't have a pk_id need to be inserted
-    insert_mask = events["pk_id"].isna()
+    events["do_insert"] = events["pk_id"].isna()
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:288]: also save `do_insert` mask

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py [hunk 7]
```diff
@@ -379,7 +372,17 @@ def process_gtfs_rt_files(db_manager: DatabaseManager) -> None:
                 tu_events = process_tu_files(files["tu_paths"], db_manager)
                 events = combine_events(vp_events, tu_events)
 
-            upload_to_database(events, db_manager)
+            # continue events processing if records exist
+            if events.shape[0] > 0:
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:376]: `upload_to_database`, `process_trips` and `process_metrics` should only be run if events exist in dataframe. We could throw logging statements to indicate this??

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py [hunk 8]
```diff
@@ -379,7 +372,17 @@ def process_gtfs_rt_files(db_manager: DatabaseManager) -> None:
...
+                events = upload_to_database(events, db_manager)
+                # drop all records that do not require update or insert
+                events = events[(events["do_update"]) | (events["do_insert"])]
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:379]: drop any events that are not related to insert or update actions

File: python_src/src/lamp_py/performance_manager/l1_cte_statements.py [hunk 9]
```diff
@@ -94,8 +95,16 @@ def get_static_trips_cte(
...
+        .join(
+            StaticRoutes,
+            sa.and_(
+                StaticStopTimes.timestamp == StaticRoutes.timestamp,
+                StaticTrips.route_id == StaticRoutes.route_id,
+            ),
+        )
         .where(
             (StaticStopTimes.timestamp.in_(timestamps))
+            & (StaticRoutes.route_type != 3)
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l1_cte_statements.py:107]: Limit static trips data to Non-BUS records.

File: python_src/src/lamp_py/performance_manager/l1_cte_statements.py [hunk 10]
```diff
@@ -116,12 +125,12 @@ def get_rt_trips_cte(start_date: int) -> sa.sql.selectable.CTE:
...
-            VehicleEvents.route_id,
-            VehicleEvents.start_date,
-            VehicleEvents.start_time,
-            VehicleEvents.vehicle_id,
+            VehicleTrips.fk_static_timestamp,
+            VehicleTrips.direction_id,
+            VehicleTrips.route_id,
+            VehicleTrips.start_date,
+            VehicleTrips.start_time,
+            VehicleTrips.vehicle_id,
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l1_cte_statements.py:133]: Pull these values from `VehicleTrips` so that they columns can be dropped from the `VehicleEvents` table.

File: python_src/src/lamp_py/performance_manager/l1_rt_trips.py [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+    """
+    insert_columns = [
+        "trip_hash",
+        "fk_static_timestamp",
+        "direction_id",
+        "route_id",
+        "start_date",
+        "start_time",
+        "vehicle_id",
+    ]
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l1_rt_trips.py:37]: This function will insert basic trips information for any new trips no currently represented in the RDS.

File: python_src/src/lamp_py/performance_manager/l1_rt_trips.py [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+    seed_timestamps: List[int],
+) -> None:
+    """
+    processs updates to trips table
+
+    generate new trips table information
+    """
+    process_logger = ProcessLogger("l1_rt_trips_table_loader")
+    process_logger.log_start()
+
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l1_rt_trips.py:102]: All of the business logic in `process_trips_table` is basically the same as before, except this will only ever result in UPDATES to the trip table and never INSERT any new trips records.

File: python_src/src/lamp_py/runtime_utils/alembic_migration.py [hunk 13]
```diff
@@ -5,22 +5,42 @@
...
+def alembic_upgrade_to_head(db_name: str) -> None:
+    """
+    upgrade db_name to head revision
+    """
     # load alembic configuation for db_name
-    alembic_cfg = Config(alembic_cfg_file, ini_section=db_name)
+    alembic_cfg = get_alembic_config(db_name)
 
-    # normal migration command, for if version table exists
     command.upgrade(alembic_cfg, revision="head")
```
Comment (rymarczy) [inline python_src/src/lamp_py/runtime_utils/alembic_migration.py:36]: more accurate function name

File: python_src/src/lamp_py/runtime_utils/alembic_migration.py [hunk 14]
```diff
@@ -5,22 +5,42 @@
...
+
+
+def alembic_downgrade_to_base(db_name: str) -> None:
+    """
+    downgrade db_name to base revision
+    """
+    # load alembic configuation for db_name
+    alembic_cfg = get_alembic_config(db_name)
+
+    command.downgrade(alembic_cfg, revision="base")
```
Comment (rymarczy) [inline python_src/src/lamp_py/runtime_utils/alembic_migration.py:46]: Use alembic to tear DB down

File: python_src/src/lamp_py/postgres/seed_metadata.py [hunk 15]
```diff
@@ -94,10 +96,8 @@ def run() -> None:
     db_manager = DatabaseManager(parsed_args.verbose)
 
     if parsed_args.clear_static:
-        SqlBase.metadata.drop_all(db_manager.engine)
-        drop_alembic_sql = "DROP TABLE IF EXISTS alembic_version;"
-        db_manager.execute(sa.text(drop_alembic_sql))
-        run_alembic_migration("performance_manager")
+        alembic_downgrade_to_base("performance_manager")
+        alembic_upgrade_to_head("performance_manager")
```
Comment (rymarczy) [inline python_src/src/lamp_py/postgres/seed_metadata.py:100]: use alembic to drop all tables from RDS, this is required to correctly move triggers from tables in proper order

File: python_src/src/lamp_py/migrations/versions/performance_manager/2e9d81949da8_trips_and_metrics_processing_updates.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+def downgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_constraint(
+        "vehicle_trips_fk_static_timestamp_fkey",
+        "vehicle_trips",
+        type_="foreignkey",
+    )
+    op.drop_column("vehicle_trips", "fk_static_timestamp")
+    op.execute("DELETE FROM vehicle_trips WHERE stop_count IS NULL;")
```
Comment (rymarczy) [inline python_src/src/lamp_py/migrations/versions/performance_manager/2e9d81949da8_trips_and_metrics_processing_updates.py:55]: NULL records must be removed from DB before converting column back to NON-NULLABLE

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py [hunk 17]
```diff
@@ -143,10 +136,6 @@ def upload_to_database(
     already in the database and those that are brand new. update preexisting
     events where appropriate and insert the new ones.
     """
-    # handle empty events dataframe
-    if events.shape[0] == 0:
-        return
-
```
Comment (rymarczy) [inline python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:149]: move to `process_gtfs_rt_files` function

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py (hunk 4 above)
Comment (mzappitello) [reply python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:250]: what does the _db stand for?

File: python_src/src/lamp_py/performance_manager/l1_rt_trips.py [hunk 18]
```diff
@@ -0,0 +1,442 @@
...
+                columns={"trip_hash": "b_trip_hash"}
+            ),
+        )
+
+    process_logger.log_complete()
+
+
+def process_trips(
+    db_manager: DatabaseManager, events: pandas.DataFrame
+) -> pandas.DataFrame:
```
Comment (mzappitello) [inline python_src/src/lamp_py/performance_manager/l1_rt_trips.py:423]: doesn't look like this function has a return statement.

File: python_src/src/lamp_py/postgres/postgres_schema.py
Comment (mzappitello) [inline python_src/src/lamp_py/postgres/postgres_schema.py:1]: are you planning on taking the vehicle trip information out of the vehicle events table eventually?

File: python_src/src/lamp_py/runtime_utils/alembic_migration.py (hunk 14 above)
Comment (mzappitello) [reply python_src/src/lamp_py/runtime_utils/alembic_migration.py:46]: is the base config just an empty db?

**Review by mzappitello — APPROVED** [review]
This looks good to me. I still find some of the ways that data moves through the biz logic strange from a design standpoint, but this does seem to work as i'd expect.

File: python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py (hunk 4 above)
Comment (rymarczy) [reply python_src/src/lamp_py/performance_manager/l0_gtfs_rt_events.py:250]: basically means "from database", these columns and the move and stop timestamps pulled from the database for comparison purposes, open to different nomenclature

File: python_src/src/lamp_py/runtime_utils/alembic_migration.py (hunk 14 above)
Comment (rymarczy) [reply python_src/src/lamp_py/runtime_utils/alembic_migration.py:46]: yes, except for the `alembic_version` table.

File: python_src/src/lamp_py/performance_manager/l1_rt_trips.py (hunk 18 above)
Comment (rymarczy) [reply python_src/src/lamp_py/performance_manager/l1_rt_trips.py:423]: What the heck, pylint is asleep at the wheel...

File: python_src/src/lamp_py/postgres/postgres_schema.py
Comment (rymarczy) [reply python_src/src/lamp_py/postgres/postgres_schema.py:1]: Yes, that is the plan, but didn't want to rock the boat too much with the OPMI tableau systems.
